spikeinterface/toolkit/qualitymetrics/pca_metrics.py

```diff
@@ -298,21 +301,152 @@ def nearest_neighbors_isolation(all_pcs, all_labels, this_unit_id, max_spikes_fo
             pcs_other_unit_idx = rng.choice(np.arange(n_spikes_other_unit), size=spikes_for_nn_actual)
             pcs_other_unit = pcs_other_unit[pcs_other_unit_idx]
 
-        pcs_concat = np.concatenate((pcs_target_unit, pcs_other_unit), axis=0)
-        label_concat = np.concatenate((np.zeros(spikes_for_nn_actual),np.ones(spikes_for_nn_actual)))
-        
-        # if n_neighbors is greater than the number of spikes in both clusters, then set it to max possible
-        if n_neighbors > len(label_concat):
-            n_neighbors_adjusted = len(label_concat)-1
-        else:
-            n_neighbors_adjusted = n_neighbors
+        isolation[other_unit_id==other_units_ids] = _compute_isolation(pcs_target_unit, pcs_other_unit, n_neighbors)
+    
+    nn_isolation = np.min(isolation)
+    return nn_isolation
+
+def nearest_neighbors_noise_overlap(waveform_extractor: si.WaveformExtractor, 
+                                    this_unit_id: int, max_spikes_for_nn: int=1000,
+                                    n_neighbors: int=5, n_components: int=10, seed: int=0):
+    """Calculates unit noise overlap based on NearestNeighbors search in PCA space.
+
+    Based on noise overlap metric described in Chung et al. (2017) Neuron 95: 1381-1394.
+
+    Rough logic:
+    ------------
+    1) Generate a noise cluster by randomly sampling voltage snippets from recording.
+    2) Subtract projection onto the weighted average of noise snippets
+       of both the target and noise clusters to correct for bias in sampling.
+    3) Compute the isolation score between the noise cluster and the target cluster.
+    
+    Implementation details:
+    -----------------------
+    As with nn_isolation, the clusters that are compared (target and noise clusters)
+    have the same number of spikes.
+    
+    See docstring for `_compute_isolation` for the definition of isolation score.
+    
+    Parameters:
+    -----------
+    we: si.WaveformExtractor
+    this_unit_id: int
+        ID of unit for which this metric will be calculated
+    max_spikes_for_nn: int
+        max number of spikes to use per cluster
+    n_neighbors: int
+        number of neighbors to check membership of
+    n_components: int
+        number of PC components to project the snippets
+    seed: int
+        seed for random subsampling of spikes
+
+    Outputs:
+    --------
+    nn_noise_overlap : float
+    """
+
+    # set random seed
+    rng = np.random.default_rng(seed=seed)
+    
+    # get random snippets from the recording to create a noise cluster
+    recording = waveform_extractor.recording
+    noise_cluster = get_random_data_chunks(recording, return_scaled=True,
```

Comment:
good point, but I think in this case the 'null hypothesis' is that the cluster was sampled randomly - if we generate the noise cluster by taking the snippets in the parts of the recording that don't have spikes, then we are essentially guaranteeing that the noise cluster won't contain any threshold crossing events (since if they did, they would have been detected as spikes) and it would be too easy for even a poorly isolated cluster pass the test.